File: backend-python/src/logic/question_engine.py
"""
Question Engine Module

Verwaltet adaptive Fragebögen für verschiedene Gewerke
"""

import os
import json
from typing import List, Dict, Optional


class QuestionEngine:
    def __init__(self, llm_provider):
        self.llm_provider = llm_provider
        self.questions_cache = {}
        self.load_questions()

    def load_questions(self):
        """Lädt vordefinierte Fragen für alle Gewerke"""
        # Sanitär-Fragen
        self.questions_cache['sanitaer'] = [
            {
                'id': 'sanitaer_1',
                'question': 'Welche Sanitärobjekte sollen installiert werden?',
                'type': 'multiple_choice',
                'options': ['WC', 'Waschbecken', 'Dusche', 'Badewanne', 'Urinal'],
                'required': True
            },
            {
                'id': 'sanitaer_2',
                'question': 'Sollen bestehende Sanitärobjekte demontiert werden?',
                'type': 'yes_no',
                'required': True
            },
            {
                'id': 'sanitaer_3',
                'question': 'Welche Art von Armaturen bevorzugen Sie?',
                'type': 'single_choice',
                'options': ['Standard', 'Hochwertig', 'Premium'],
                'required': False
            }
        ]
        
        # Fliesen-Fragen
        self.questions_cache['fliesen'] = [
            {
                'id': 'fliesen_1',
                'question': 'Welche Fläche soll gefliest werden (in m²)?',
                'type': 'number',
                'required': True
            },
            {
                'id': 'fliesen_2',
                'question': 'Welche Art von Fliesen bevorzugen Sie?',
                'type': 'single_choice',
                'options': ['Keramik', 'Naturstein', 'Feinsteinzeug'],
                'required': True
            },
            {
                'id': 'fliesen_3',
                'question': 'Sollen alte Fliesen entfernt werden?',
                'type': 'yes_no',
                'required': True
            },
            {
                'id': 'fliesen_4',
                'question': 'Welche Fliesengröße bevorzugen Sie?',
                'type': 'single_choice',
                'options': ['Klein (bis 30x30cm)', 'Mittel (30x60cm)', 'Groß (60x60cm oder größer)'],
                'required': False
            },
            {
                'id': 'fliesen_5',
                'question': 'Benötigen Sie eine Fußbodenheizung?',
                'type': 'yes_no',
                'required': False
            }
        ]
        
        # Maler-Fragen
        self.questions_cache['maler'] = [
            {
                'id': 'maler_1',
                'question': 'Welche Fläche soll gestrichen werden (in m²)?',
                'type': 'number',
                'required': True
            },
            {
                'id': 'maler_2',
                'question': 'Welche Art von Anstrich bevorzugen Sie?',
                'type': 'single_choice',
                'options': ['Dispersionsfarbe', 'Latexfarbe', 'Silikatfarbe'],
                'required': True
            },
            {
                'id': 'maler_3',
                'question': 'Benötigen Sie Vorarbeiten (Spachteln, Grundierung)?',
                'type': 'yes_no',
                'required': True
            }
        ]

    def detect_trades(self, description: str) -> List[str]:
        """Erkennt Gewerke aus der Projektbeschreibung"""
        description_lower = description.lower()
        detected_trades = []
        
        # Keyword-basierte Erkennung
        trade_keywords = {
            'sanitaer': ['sanitär', 'bad', 'wc', 'dusche', 'badewanne', 'waschbecken', 'toilette'],
            'fliesen': ['fliesen', 'kacheln', 'fliesenleger', 'bodenfliesen', 'wandfliesen'],
            'maler': ['maler', 'streichen', 'farbe', 'anstrich', 'tapete', 'malerei'],
            'elektro': ['elektro', 'strom', 'licht', 'schalter', 'steckdose', 'elektrik'],
            'heizung': ['heizung', 'heizkörper', 'thermostat', 'warmwasser'],
            'tischler': ['tischler', 'schreiner', 'möbel', 'schrank', 'holz'],
            'trockenbau': ['trockenbau', 'rigips', 'gipskarton', 'wand', 'decke']
        }
        
        for trade, keywords in trade_keywords.items():
            if any(keyword in description_lower for keyword in keywords):
                detected_trades.append(trade)
        
        # Fallback: Wenn nichts erkannt wurde, verwende Standard-Gewerke
        if not detected_trades:
            detected_trades = ['sanitaer', 'fliesen', 'maler']
        
        return detected_trades

    def get_questions_for_trade(self, trade: str) -> List[Dict]:
        """Gibt Fragen für ein bestimmtes Gewerk zurück"""
        return self.questions_cache.get(trade, [])

    def validate_answers(self, trade: str, answers: Dict) -> Dict:
        """Validiert Antworten für ein Gewerk"""
        questions = self.get_questions_for_trade(trade)
        validation_result = {
            'valid': True,
            'errors': [],
            'warnings': []
        }
        
        for question in questions:
            question_id = question['id']
            answer = answers.get(question_id)
            
            # Prüfe erforderliche Fragen
            if question.get('required', False) and not answer:
                validation_result['valid'] = False
                validation_result['errors'].append(f"Frage '{question['question']}' ist erforderlich")
                continue
            
            # Typ-spezifische Validierung
            if answer:
                if question['type'] == 'number':
                    try:
                        float(answer)
                    except (ValueError, TypeError):
                        validation_result['valid'] = False
                        validation_result['errors'].append(f"'{answer}' ist keine gültige Zahl")
                
                elif question['type'] in ['single_choice', 'multiple_choice']:
                    options = question.get('options', [])
                    if question['type'] == 'single_choice':
                        if answer not in options:
                            validation_result['valid'] = False
                            validation_result['errors'].append(f"'{answer}' ist keine gültige Option")
                    elif question['type'] == 'multiple_choice':
                        if isinstance(answer, list):
                            for item in answer:
                                if item not in options:
                                    validation_result['valid'] = False
                                    validation_result['errors'].append(f"'{item}' ist keine gültige Option")
                        else:
                            validation_result['valid'] = False
                            validation_result['errors'].append("Mehrfachauswahl erwartet eine Liste")
                
                elif question['type'] == 'yes_no':
                    if answer not in ['yes', 'no', 'ja', 'nein', True, False]:
                        validation_result['valid'] = False
                        validation_result['errors'].append(f"'{answer}' ist keine gültige Ja/Nein-Antwort")
        
        return validation_result

    def get_trade_summary(self, trade: str, answers: Dict) -> str:
        """Erstellt eine Zusammenfassung der Antworten für ein Gewerk"""
        questions = self.get_questions_for_trade(trade)
        summary_parts = []
        
        for question in questions:
            question_id = question['id']
            answer = answers.get(question_id)
            
            if answer:
                summary_parts.append(f"- {question['question']}: {answer}")
        
        return f"Zusammenfassung {trade.upper()}:\n" + "\n".join(summary_parts)

File: backend-python/src/logic/question_engine.js
/**
 * Adaptive Fragenfluss-Engine
 * 
 * Verwaltet den dynamischen Fragenfluss basierend auf Projektart und Gewerk
 */

const fs = require('fs-extra');
const path = require('path');
const llmProvider = require('../llm/provider');

class QuestionEngine {
  constructor() {
    this.sessions = new Map();
    this.promptsDir = path.join(__dirname, '../../../prompts');
    this.questionCache = new Map();
  }

  /**
   * Erstellt eine neue Fragen-Session
   */
  async createSession(projectData) {
    const sessionId = this.generateSessionId();
    
    const session = {
      id: sessionId,
      projectData,
      currentTrade: null,
      currentStep: 0,
      answers: {},
      tradeProgress: {},
      detectedTrades: [],
      createdAt: new Date(),
      lastActivity: new Date(),
    };

    // Erkenne relevante Gewerke basierend auf Projektbeschreibung
    session.detectedTrades = await this.detectTrades(projectData);
    
    // Initialisiere Progress für jedes Gewerk
    session.detectedTrades.forEach(trade => {
      session.tradeProgress[trade] = {
        currentQuestion: 0,
        totalQuestions: 0,
        completed: false,
        questions: [],
        answers: {},
      };
    });

    this.sessions.set(sessionId, session);
    return session;
  }

  /**
   * Erkennt relevante Gewerke basierend auf Projektbeschreibung
   */
  async detectTrades(projectData) {
    const { category, subCategory, description } = projectData;
    
    // Verwende LLM für intelligente Gewerke-Erkennung
    if (llmProvider.isAvailable()) {
      try {
        const prompt = `
Analysiere das folgende Bauprojekt und bestimme die relevanten Gewerke:

Kategorie: ${category}
Unterkategorie: ${subCategory || 'Nicht angegeben'}
Beschreibung: ${description}

Verfügbare Gewerke:
- sanitaer (Sanitärinstallation)
- elektro (Elektroinstallation)
- heizung (Heizung/Lüftung)
- fliesen (Fliesenarbeiten)
- maler (Malerarbeiten)
- trockenbau (Trockenbauarbeiten)
- dachdecker (Dacharbeiten)
- fenster-tueren (Fenster/Türen)
- fassadenbau (Fassadenarbeiten)
- geruest (Gerüstbau)
- aussenanlagen (Außenanlagen)
- tischler (Tischlerarbeiten)
- schlosser (Schlosserarbeiten)

Antworte nur mit einer kommagetrennten Liste der relevanten Gewerke-IDs.
`;

        const response = await llmProvider.chatComplete({
          messages: [{ role: 'user', content: prompt }],
          temperature: 0.3,
        });

        const trades = response.split(',').map(t => t.trim()).filter(t => t);
        return trades.length > 0 ? trades : this.fallbackTradeDetection(description);
        
      } catch (error) {
        console.warn('⚠️ LLM Gewerke-Erkennung fehlgeschlagen, verwende Fallback:', error.message);
      }
    }

    return this.fallbackTradeDetection(description);
  }

  /**
   * Fallback Gewerke-Erkennung basierend auf Keywords
   */
  fallbackTradeDetection(description) {
    const lower = description.toLowerCase();
    const trades = [];

    const tradeKeywords = {
      sanitaer: ['bad', 'sanitär', 'wc', 'dusche', 'badewanne', 'waschbecken', 'toilette'],
      elektro: ['elektro', 'strom', 'licht', 'steckdose', 'schalter', 'beleuchtung'],
      heizung: ['heizung', 'heizkörper', 'thermostat', 'warmwasser', 'boiler'],
      fliesen: ['fliesen', 'kacheln', 'bodenfliesen', 'wandfliesen'],
      maler: ['maler', 'streichen', 'tapete', 'farbe', 'anstrich'],
      trockenbau: ['trockenbau', 'rigips', 'gipskarton', 'wand', 'decke'],
      dachdecker: ['dach', 'ziegel', 'dachrinne', 'schornstein'],
      'fenster-tueren': ['fenster', 'tür', 'türen', 'rolladen'],
      fassadenbau: ['fassade', 'außenwand', 'dämmung', 'putz'],
    };

    Object.entries(tradeKeywords).forEach(([trade, keywords]) => {
      if (keywords.some(keyword => lower.includes(keyword))) {
        trades.push(trade);
      }
    });

    // Mindestens ein Gewerk zurückgeben
    return trades.length > 0 ? trades : ['sanitaer'];
  }

  /**
   * Lädt Fragen für ein spezifisches Gewerk
   */
  async loadTradeQuestions(tradeName) {
    const cacheKey = `questions_${tradeName}`;
    
    if (this.questionCache.has(cacheKey)) {
      return this.questionCache.get(cacheKey);
    }

    try {
      const promptFile = `${tradeName}_fragenprompt_optimiert.md`;
      const promptPath = path.join(this.promptsDir, promptFile);
      
      if (!(await fs.pathExists(promptPath))) {
        console.warn(`⚠️ Fragenprompt nicht gefunden: ${promptFile}`);
        return [];
      }

      const content = await fs.readFile(promptPath, 'utf8');
      const questions = this.parseQuestionsFromPrompt(content, tradeName);
      
      this.questionCache.set(cacheKey, questions);
      return questions;
      
    } catch (error) {
      console.error(`❌ Fehler beim Laden der Fragen für ${tradeName}:`, error);
      return [];
    }
  }

  /**
   * Parst Fragen aus einem Fragenprompt
   */
  parseQuestionsFromPrompt(content, tradeName) {
    const questions = [];
    const lines = content.split('\n');
    
    let currentSection = null;
    let currentQuestion = null;
    let questionCounter = 1;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i].trim();
      
      // Erkenne Abschnitte
      if (line.startsWith('## ') || line.startsWith('### ')) {
        currentSection = line.replace(/^#+\s*/, '');
        continue;
      }

      // Erkenne Fragen (verschiedene Formate)
      const questionPatterns = [
        /^\*\*(.+?)\*\*\s*\*\((.+?)\)\*/,  // **Frage** *(Erklärung)*
        /^\*\*(.+?)\*\*/,                   // **Frage**
        /^(.+?)\?\s*\*\((.+?)\)\*/,        // Frage? *(Erklärung)*
        /^(.+?)\?$/,                       // Frage?
      ];

      for (const pattern of questionPatterns) {
        const match = line.match(pattern);
        if (match) {
          const questionText = match[1].trim();
          const explanation = match[2] || null;
          
          // Sammle Antwortoptionen aus den folgenden Zeilen
          const options = this.extractOptionsFromFollowingLines(lines, i + 1);
          
          const question = {
            id: `${tradeName}_q${questionCounter}`,
            trade: tradeName,
            section: currentSection,
            text: questionText,
            explanation,
            type: this.determineQuestionType(questionText, options),
            options: options.length > 0 ? options : null,
            required: this.isRequiredQuestion(questionText),
            dependencies: this.extractDependencies(questionText),
          };

          questions.push(question);
          questionCounter++;
          break;
        }
      }
    }

    return questions;
  }

  /**
   * Extrahiert Antwortoptionen aus folgenden Zeilen
   */
  extractOptionsFromFollowingLines(lines, startIndex) {
    const options = [];
    
    for (let i = startIndex; i < lines.length && i < startIndex + 10; i++) {
      const line = lines[i].trim();
      
      // Stoppe bei neuer Frage oder Abschnitt
      if (line.startsWith('**') || line.startsWith('##') || line === '') {
        break;
      }

      // Erkenne Optionen
      const optionPatterns = [
        /^-\s*🔴\s*\*\*(.+?)\*\*(.*)$/,     // - 🔴 **Option** Beschreibung
        /^-\s*🟡\s*\*\*(.+?)\*\*(.*)$/,     // - 🟡 **Option** Beschreibung
        /^-\s*🟢\s*\*\*(.+?)\*\*(.*)$/,     // - 🟢 **Option** Beschreibung
        /^-\s*\*\*(.+?)\*\*(.*)$/,          // - **Option** Beschreibung
        /^-\s*(.+?)$/,                      // - Option
      ];

      for (const pattern of optionPatterns) {
        const match = line.match(pattern);
        if (match) {
          const value = match[1].trim();
          const description = match[2] ? match[2].trim() : '';
          
          options.push({
            value,
            label: value,
            description: description.replace(/^-\s*/, ''),
          });
          break;
        }
      }
    }

    return options;
  }

  /**
   * Bestimmt den Fragentyp basierend auf Inhalt
   */
  determineQuestionType(questionText, options) {
    if (options && options.length > 0) {
      return options.length > 5 ? 'select' : 'radio';
    }
    
    if (questionText.toLowerCase().includes('größe') || 
        questionText.toLowerCase().includes('abmessungen') ||
        questionText.toLowerCase().includes('meter')) {
      return 'number';
    }
    
    if (questionText.toLowerCase().includes('beschreib') ||
        questionText.toLowerCase().includes('erläuter')) {
      return 'textarea';
    }
    
    return 'text';
  }

  /**
   * Prüft, ob eine Frage erforderlich ist
   */
  isRequiredQuestion(questionText) {
    const requiredKeywords = ['wichtig', 'erforderlich', 'muss', 'grunddaten'];
    return requiredKeywords.some(keyword => 
      questionText.toLowerCase().includes(keyword)
    );
  }

  /**
   * Extrahiert Abhängigkeiten zwischen Fragen
   */
  extractDependencies(questionText) {
    // Vereinfachte Abhängigkeits-Erkennung
    // In einer vollständigen Implementierung würde dies komplexer sein
    return [];
  }

  /**
   * Gibt die nächsten Fragen für eine Session zurück
   */
  async getNextQuestions(sessionId, tradeId = null) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session nicht gefunden');
    }

    // Aktualisiere letzte Aktivität
    session.lastActivity = new Date();

    // Wenn kein spezifisches Gewerk angegeben, nimm das erste unvollständige
    if (!tradeId) {
      tradeId = this.getNextIncompleteTrade(session);
    }

    if (!tradeId) {
      return { completed: true, questions: [] };
    }

    // Lade Fragen für das Gewerk falls noch nicht geschehen
    if (session.tradeProgress[tradeId].questions.length === 0) {
      const questions = await this.loadTradeQuestions(tradeId);
      session.tradeProgress[tradeId].questions = questions;
      session.tradeProgress[tradeId].totalQuestions = questions.length;
    }

    const tradeProgress = session.tradeProgress[tradeId];
    const currentIndex = tradeProgress.currentQuestion;
    
    // Gib nächste 3-5 Fragen zurück (adaptiv)
    const batchSize = this.calculateBatchSize(tradeId, currentIndex);
    const nextQuestions = tradeProgress.questions.slice(currentIndex, currentIndex + batchSize);

    return {
      sessionId,
      tradeId,
      tradeName: tradeId,
      currentStep: currentIndex + 1,
      totalSteps: tradeProgress.totalQuestions,
      questions: nextQuestions,
      progress: Math.round((currentIndex / tradeProgress.totalQuestions) * 100),
      completed: false,
    };
  }

  /**
   * Berechnet die optimale Batch-Größe für Fragen
   */
  calculateBatchSize(tradeId, currentIndex) {
    // Komplexe Gewerke: kleinere Batches für bessere UX
    const complexTrades = ['elektro', 'heizung', 'sanitaer'];
    
    if (complexTrades.includes(tradeId)) {
      return currentIndex === 0 ? 3 : 2; // Erste Fragen: 3, dann 2
    }
    
    return currentIndex === 0 ? 5 : 3; // Einfache Gewerke: größere Batches
  }

  /**
   * Findet das nächste unvollständige Gewerk
   */
  getNextIncompleteTrade(session) {
    for (const trade of session.detectedTrades) {
      if (!session.tradeProgress[trade].completed) {
        return trade;
      }
    }
    return null;
  }

  /**
   * Speichert Antworten für eine Session
   */
  async saveAnswers(sessionId, answers) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session nicht gefunden');
    }

    session.lastActivity = new Date();

    // Speichere Antworten
    Object.entries(answers).forEach(([questionId, answer]) => {
      session.answers[questionId] = {
        questionId,
        answer: answer.value || answer,
        assumption: answer.assumption || null,
        timestamp: new Date(),
      };

      // Aktualisiere Trade-Progress
      const trade = questionId.split('_')[0];
      if (session.tradeProgress[trade]) {
        session.tradeProgress[trade].answers[questionId] = answer;
        session.tradeProgress[trade].currentQuestion++;
      }
    });

    // Prüfe, ob Gewerk abgeschlossen ist
    this.checkTradeCompletion(session);

    return {
      success: true,
      nextAvailable: this.getNextIncompleteTrade(session) !== null,
    };
  }

  /**
   * Prüft, ob ein Gewerk abgeschlossen ist
   */
  checkTradeCompletion(session) {
    Object.entries(session.tradeProgress).forEach(([trade, progress]) => {
      if (progress.currentQuestion >= progress.totalQuestions && !progress.completed) {
        progress.completed = true;
        console.log(`✅ Gewerk ${trade} abgeschlossen`);
      }
    });
  }

  /**
   * Gibt Session-Informationen zurück
   */
  getSessionInfo(sessionId) {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error('Session nicht gefunden');
    }

    const totalQuestions = Object.values(session.tradeProgress)
      .reduce((sum, progress) => sum + progress.totalQuestions, 0);
    
    const answeredQuestions = Object.keys(session.answers).length;
    
    const completedTrades = Object.values(session.tradeProgress)
      .filter(progress => progress.completed).length;

    return {
      sessionId: session.id,
      projectData: session.projectData,
      detectedTrades: session.detectedTrades,
      totalTrades: session.detectedTrades.length,
      completedTrades,
      totalQuestions,
      answeredQuestions,
      overallProgress: totalQuestions > 0 ? Math.round((answeredQuestions / totalQuestions) * 100) : 0,
      tradeProgress: session.tradeProgress,
      createdAt: session.createdAt,
      lastActivity: session.lastActivity,
    };
  }

  /**
   * Generiert eine eindeutige Session-ID
   */
  generateSessionId() {
    return `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Bereinigt alte Sessions (älter als 24 Stunden)
   */
  cleanupOldSessions() {
    const cutoff = new Date(Date.now() - 24 * 60 * 60 * 1000);
    
    for (const [sessionId, session] of this.sessions.entries()) {
      if (session.lastActivity < cutoff) {
        this.sessions.delete(sessionId);
        console.log(`🗑️ Session ${sessionId} bereinigt (inaktiv)`);
      }
    }
  }
}

// Singleton-Instanz
const questionEngine = new QuestionEngine();

// Automatische Bereinigung alle 6 Stunden
setInterval(() => {
  questionEngine.cleanupOldSessions();
}, 6 * 60 * 60 * 1000);

module.exports = questionEngine;

